src/actions.sol

```diff
@@ -58,8 +66,31 @@ contract Actions {
     event ApproveERC20(address indexed erc20, address indexed usr, uint256 amount);
     event TransferERC20(address indexed erc20, address indexed dst, uint256 amount);
 
+    address public immutable shelf;
+    address public immutable pile;
+    address public immutable self;
+
+    // address to deposit withdraws
+    address public immutable withdrawAddress;
+
+    // modifier only delegated call
+    modifier onlyDelegateCall() {
+        require(address(this) != self, "only-delegate-call");
+        _;
+    }
+
+    constructor(address root_, address withdrawAddress_) {
+        shelf = BorrowerDeployerLike(RootLike(root_).borrowerDeployer()).shelf();
+        pile = ShelfLike(shelf).pile();
+        self = address(this);
+        require(withdrawAddress_ != address(0), "withdraw-address-not-set");
+        require(shelf != address(0), "shelf-not-set");
+        require(pile != address(0), "pile-not-set");
+        withdrawAddress = withdrawAddress_;
+    }
+
     // --- Borrower Actions ---
-    function issue(address shelf, address registry, uint256 token) public returns (uint256 loan) {
+    function issue(address registry, uint256 token) public onlyDelegateCall returns (uint256 loan) {
```

Comment:
Just a general comment on most of these methods: I think we should consider keeping all the `address` parameters for shelf and pile, and just reverting if the passed param doesn't match the stored one. Otherwise, the UI will need to migrate every method call, and more importantly, keep logic for both old and new versions of the actions.